AG-173 - Make local cache customizable

barreiro opened the pull request.

jesperpedersen commented:
I would have the `Acquirable` and a `ConnectionCache` interface as the main interfaces for this feature.

Then have `LocalConnectionCache` extend `ConnectionCache` (or implement) which people can use. Or supply their own implementation. Mixing it - as it is now - is a bit confusing for external providers.

Where does `.fixed(4)` come from ? And, why is `.fixed()` needed ? Apps would expect `.single()` or `.full()`, no ?

barreiro commented:
> Where does `.fixed(4)` come from ? And, why is `.fixed()` needed ? Apps would expect `.single()` or `.full()`, no ?

`fixed(4)` is the current implementation. will change this default after this commit is merged.

jesperpedersen commented:
Good to go, as you said you wanted to fix the `.fixed()` case after

stuartwdouglas commented on agroal-api/src/main/java/io/agroal/api/cache/LocalConnectionCache.java:
```diff
@@ -0,0 +1,137 @@
+
+            @Override
+            public void put(Acquirable acquirable) {
+                Acquirable[] cacheArray = threadLocal.get();
```
Why does the one above have an isAcquirable() check but not this one?

barreiro replied:
yes, it's missing. Pool calls put on all connection return even when these are to be flushed.

stuartwdouglas commented on agroal-api/src/main/java/io/agroal/api/cache/LocalConnectionCache.java:
```diff
@@ -0,0 +1,137 @@
+            public void put(Acquirable acquirable) {
+                Acquirable[] cacheArray = threadLocal.get();
+                int last = cacheArray.length - 1;
+                if ( cacheArray[last] != null ) { // replace last
```
What is the reasoning behind the logic here? It fills the last spot first, then fills from the front, but aquires from the back?

If you have two items in here this means you get one at the front and one at the back? If there is a reason for this can you add a comment explaining why?

barreiro replied:
the logic is that the latest put has always preference, as it will be likely that can be acquired again. the other slots are remembered in the hope that whoever acquired it had released it between. 

anyway, this implementation is to mimic the existing behavior (for benchmarking purposes) and the plan is to change the default behavior (in favor of one of the other implementations, likely `single()`).

barreiro replied:
changed the insert order to fill from last to first.

stuartwdouglas commented on agroal-api/src/main/java/io/agroal/api/cache/ConnectionCache.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
+ */
+public interface ConnectionCache {
```
Shouldn't this have <T extends Acquirable> and get/put take type T so there is no need to cast?

barreiro replied:
There will always be a need to cast, as the `ConnectionHandler` class is not on the API.

stuartwdouglas commented on agroal-api/src/main/java/io/agroal/api/cache/LocalConnectionCache.java:
```diff
@@ -0,0 +1,137 @@
+
+            @Override
+            public void put(Acquirable acquirable) {
+                threadLocal.get().addFirst( acquirable );
```
Shouldn't this have an isAcquirable check?

barreiro replied:
yes, this needs to be fixed.

stuartwdouglas reviewed (changes requested):
This should be ok for our needs in Quarkus, but it looks like there are a few other minor issues
